02_knight_and_princess/alexey_mesheryakov/graph.cpp

```diff
@@ -1,76 +1,85 @@
+#include <assert.h>
 #include <fstream>
-#include <iostream>
 #include <sstream>
-#include <stdexcept>
 #include <string>
-#include <utility>
 #include <vector>
 
-constexpr int LEN = 14;
+constexpr int VERTEX_CNT = 14;
 
-using std::cout;
 using std::pair;
 using std::vector;
 using EdgeId = int;
 using VertexId = int;
 
 class Graph {
  public:
-  struct Edge {
-    EdgeId id = 0;
-    pair<VertexId, VertexId> vertex = {-1, -1};
-  };
+  class Edge {
+   public:
+    Edge(const pair<VertexId, VertexId>& vertex_ids, const EdgeId& edge_id)
+        : vertex_ids_(vertex_ids), id_(edge_id) {}
+
+    std::string to_string() const {
+      std::stringstream buffer;
+      buffer << "{\"id\":" << id_ << ",\"vertex_ids\":[" << vertex_ids_.first
+             << "," << vertex_ids_.second << "]}";
+      return buffer.str();
+    }
 
-  struct Vertex {
-    int id = -1;
-    vector<EdgeId> edges = {};
+   private:
+    EdgeId id_ = 0;
+    pair<VertexId, VertexId> vertex_ids_ = {-1, -1};
   };
 
-  void add_vertex(const Vertex& v) { vertex_ids_.push_back(v); }
+  class Vertex {
+   public:
+    Vertex(const VertexId& id) : id_(id) {}
 
-  void add_edge(Edge edge) {
-    edges_.push_back(edge);
-    try {
-      vertex_ids_[edge.vertex.first].edges.push_back(edge.id);
-      vertex_ids_[edge.vertex.second].edges.push_back(edge.id);
-    } catch (std::out_of_range) {
-      cout << "ERROR: id does'n exist\n";
+    std::string to_string() const {
+      std::stringstream buffer;
+      buffer << "{\"id\":" << id_ << ",\"edge_ids\":[";
+      for (int i = 0; i < edges_.size() - 1; i++)
+        buffer << edges_[i] << ",";
+      buffer << edges_[edges_.size() - 1] << "]}";
+      return buffer.str();
     }
-  }
 
-  void set_params(const vector<Edge>& new_edges = {},
-                  const vector<int>& vertex_ids = {}) {
-    for (const auto& id : vertex_ids)
-      add_vertex({id, {}});
-    for (const auto& edge : new_edges)
-      add_edge(edge);
+    void add_edge(const EdgeId& edge_id) { edges_.push_back(edge_id); }
+
+   private:
+    VertexId id_ = -1;
+    vector<EdgeId> edges_ = {};
+  };
+
+  bool vertex_exist(const VertexId& id) {
+    return (0 <= id) && (id < vertex_id_counter_);
   }
 
-  std::string to_string() const {
-    std::stringstream buffer;
-    for (auto edge : edges_)
-      buffer << edge.id << " " << vertex_ids_[edge.vertex.first].id << " "
-             << vertex_ids_[edge.vertex.second].id << "\n";
-    return buffer.str();
+  void add_vertex() { vertices_.push_back(Vertex(vertex_id_counter_++)); }
```

Comment:
Для добавления в коллекцию объекта, который нужно сконструировать, есть специальная функция `emplace`, она как конструктор, принимает список аргументов для создания объекта и добавляет его в коллекцию:
```suggestion
  void add_vertex() { vertices_.emplace_back(vertex_id_counter_++); }
```